gobblin-yarn/src/main/java/org/apache/gobblin/yarn/YarnAutoScalingManager.java

```diff
@@ -0,0 +1,158 @@
+
+  @Override
+  protected void startUp() throws Exception {
+    int scheduleInterval = ConfigUtils.getInt(this.config, AUTO_SCALING_POLLING_INTERVAL_SECS,
+        DEFAULT_AUTO_SCALING_POLLING_INTERVAL_SECS);
+    log.info("Starting the " + YarnAutoScalingManager.class.getSimpleName());
+    log.info("Scheduling the auto scaling task with an interval of {} seconds", scheduleInterval);
+
+    this.autoScalingExecutor.scheduleAtFixedRate(new YarnAutoScalingRunnable(new TaskDriver(this.helixManager),
+            this.yarnService, this.partitionsPerContainer), 0,
+        scheduleInterval, TimeUnit.SECONDS);
+  }
+
+  @Override
+  protected void shutDown() throws Exception {
+    log.info("Stopping the " + YarnAutoScalingManager.class.getSimpleName());
+
+    ExecutorsUtils.shutdownExecutorService(this.autoScalingExecutor, Optional.of(log));
+  }
+
+  /**
+   * A {@link Runnable} that figures out the number of containers required for the workload
+   * and requests those containers.
+   */
+  @VisibleForTesting
+  @AllArgsConstructor
+  static class YarnAutoScalingRunnable implements Runnable {
+    private final TaskDriver taskDriver;
+    private final YarnService yarnService;
+    private final int partitionsPerContainer;
+
+    /**
+     * Iterate through the workflows configured in Helix to figure out the number of required partitions
+     * and request the {@link YarnService} to scale to the desired number of containers.
+     */
+    @Override
+    public void run() {
+      Set<String> inUseInstances = new HashSet<>();
+
+      int numPartitions = 0;
+      for (Map.Entry<String, WorkflowConfig> workFlowEntry : taskDriver.getWorkflows().entrySet()) {
+        WorkflowContext workflowContext = taskDriver.getWorkflowContext(workFlowEntry.getKey());
+
+        // Only allocate for active workflows
+        if (workflowContext == null || !workflowContext.getWorkflowState().equals(TaskState.IN_PROGRESS)) {
+          continue;
+        }
+
+        log.debug("Workflow name {} config {} context {}", workFlowEntry.getKey(), workFlowEntry.getValue(),
+            workflowContext);
+
+        WorkflowConfig workflowConfig = workFlowEntry.getValue();
+        JobDag jobDag = workflowConfig.getJobDag();
+
+        Set<String> jobs = jobDag.getAllNodes();
+
+        // sum up the number of partitions
+        for (String jobName : jobs) {
+          JobContext jobContext = taskDriver.getJobContext(jobName);
+
+          if (jobContext != null) {
+            log.debug("JobContext {} num partitions {}", jobContext, jobContext.getPartitionSet().size());
+
+            inUseInstances.addAll(jobContext.getPartitionSet().stream().map(jobContext::getAssignedParticipant)
+                .filter(e -> e != null).collect(Collectors.toSet()));
+
+            numPartitions += jobContext.getPartitionSet().size();
+          }
+        }
+      }
+
+      int numTargetContainers = (numPartitions + (this.partitionsPerContainer - 1)) / this.partitionsPerContainer;
```

Comment:
why is (this.partitionsPerContainer - 1) within parenthesis.

Reply:
Might want to document that you're basically doing ceiling(numPartitions/this.partitionsPerContainer) efficiently :)